NebulaPatcher/Patches/Dynamic/EnemyUnitComponent_Patch.cs

```diff
@@ -0,0 +1,400 @@
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private static void PlayerSearchLogic(ref EnemyUnitComponent @this, ref EnemyData enemy,
+        EnemyDFHiveSystem hive, int hatred_max_sense, float sensorRange)
+    {
+        var starId = hive.starData.id;
+        // Only search if there are players in space
+        if (!Multiplayer.Session.Combat.ActivedStarsMechaInSpace.Contains(starId)) return;
+
+        SpaceSector sector = hive.sector;
+        var sqrSensorRange = sensorRange * sensorRange;
+        var reachRange = sensorRange * 1.45f + sector.skillSystem.playerEnergyShieldRadius;
+
+        // Convert enemy local pos to upos
+        sector.TransformFromAstro_ref(enemy.astroId, out var enemyUpos, ref enemy.pos);
+
+        // Find the closest alive player to the unit
+        var closestPlayerId = -1;
+        var cloestSqrDist = float.MaxValue;
+        var players = Multiplayer.Session.Combat.Players;
+        for (var i = 0; i < players.Length; i++)
+        {
+            ref var playerPosition = ref players[i];
+            if (playerPosition.starId != starId || !playerPosition.isAlive) continue;
+            var sqrDist = Vector3.SqrMagnitude(playerPosition.skillTargetU - enemyUpos);
+            if (sqrDist < cloestSqrDist)
+            {
+                closestPlayerId = playerPosition.id;
+                cloestSqrDist = sqrDist;
+            }
+        }
+
+        if (closestPlayerId == -1) // No alive player in the system
+        {
+            if (@this.hatred.max.targetType == ETargetType.Player)
+            {
+                @this.hatred.ClearMax();
+            }
+            return;
+        }
+
+        if (enemy.astroId == hive.hiveAstroId) // Enemy is in the hive
+        {
+            var sensorCoef = ((@this.hatred.max.targetType == ETargetType.Player) ? 1f : 0.64f);
+            if (cloestSqrDist > sqrSensorRange * sensorCoef)
+            {
+                return;
+            }
+            var closestDist = (float)Math.Sqrt(cloestSqrDist);
+            var hateValue = sensorRange - closestDist;
+            if (@this.protoId == 8113) // Lancer
+            {
+                if (cloestSqrDist > 400000000.0)
+                {
+                    return;
+                }
+            }
+            else if (@this.protoId == 8112) // Humpback
+            {
+                hateValue *= 1f - Mathf.Clamp01(closestDist / 5500f);
+            }
+            if (hateValue > 0.0)
+            {
+                @this.hatred.HateTarget(ETargetType.Player, closestPlayerId, (int)hateValue, hatred_max_sense, EHatredOperation.Add);
+                return;
+            }
+        }
+        else if (enemy.astroId == hive.starData.astroId) // Enemy is in the system of the hive
+        {
+            if (cloestSqrDist > reachRange * reachRange)
+            {
+                return;
+            }
```

Comment:
Minor nitpick, could be simplified to:
```suggestion
            if (cloestSqrDist > reachRange * reachRange) return;
```